lib/charms/vault_k8s/v0/vault_autounseal.py

```diff
@@ -0,0 +1,371 @@
+    """Events raised by the vault-autounseal relation on the requirer side."""
+
+    vault_autounseal_details_ready = ops.EventSource(VaultAutounsealDetailsReadyEvent)
+
+
+class VaultAutounsealProvides(ops.Object):
+    """Manages the vault-autounseal relation from the provider side."""
+
+    on: VaultAutounsealProvidesEvents = VaultAutounsealProvidesEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+        self.framework.observe(
+            self.charm.on[relation_name].relation_created, self._on_relation_created
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken, self._on_relation_broken
+        )
+
+    def _on_relation_created(self, event: ops.RelationCreatedEvent) -> None:
+        self.on.vault_autounseal_initialize.emit(relation=event.relation)
+
+    def _on_relation_broken(self, event: ops.RelationBrokenEvent) -> None:
+        self.on.vault_autounseal_destroy.emit(relation=event.relation)
+
+    def set_vault_url(self, relation: ops.Relation, address: str) -> None:
+        """Set the vault url address in the relation databag.
+
+        Args:
+            relation: The relation for which to set the vault url.
+            address: The address of the vault server.
+
+        """
+        if not self.charm.unit.is_leader():
+            logger.warning(
+                "Attempting to set the vault url without being the leader. Ignoring the request."
+            )
+            return
+        relation.data[self.charm.app].update({"address": address})
+
+    def set_credentials_secret_id(self, relation: ops.Relation, secret_id: str) -> None:
+        """Set the credentials secret id in the relation databag.
+
+        Args:
+            relation: The relation for which to set the credentials secret id.
+            secret_id: The secret id of the Juju secret which stores the credentials for authenticating with the Vault server.
+        """
+        if relation.app is None:
+            raise ValueError("The `relation.app` is not set")
+        relation.data[self.charm.app].update({"credentials_secret_id": secret_id})
+
+    def set_ca_certificate(self, relation: ops.Relation, ca_certificate: str) -> None:
+        """Set the ca_certificate on the relation.
+
+        Args:
+            relation: The relation for which to set the ca_certificate.
+            ca_certificate: The CA certificate to use when validating the Vault server's certificate.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        if not relation:
+            logger.warning("Relation is None")
+            return
+        if not relation.active:
+            logger.warning("Relation is not active")
+            return
+        relation.data[self.charm.app]["ca_certificate"] = ca_certificate
+
```

Comment:
We should probably have a `get_outstanding_transit_requests` or something along those lines. For instance if 3 relations gets created before Vault A is up, no event will inform it that it has those outstanding requests to satisfy.

On configure, Vault A will likely have to call this method, and go through those outstanding requests and satisfy those.

Reply:
Wouldn't they create 3 `relation_created` events which create 3 `vault_autounseal_initialize` events in the charm?

Not that it changes the fact that if one of these fails there is no path to recovery (without defer), which there should be.

Reply:
> Wouldn't they create 3 relation_created events which create 3 vault_autounseal_initialize events in the charm?

That's what I thought would happen. I thought these events were promised to fire by the Juju framework. 

> Not that it changes the fact that if one of these fails there is no path to recovery (without defer), which there should be.

Using defer was my plan, here. It seems like a case where it makes sense. 

I really don't like the implication of trying to catch _every_ scenario of un-fired events in `_configure`.. but if that's the direction we're going, I'll reluctantly add it :laughing:

Reply:
If it makes sense to use defer here, it makes sense to use it in every relation that Vault provides. If we want to change the direction for defers we should be intentional about it, spec it,  and convince the team and me that it is a better idea than the current pattern. 

For the time being I would encourage following the same pattern we have for other interfaces.

Reply:
I'm open to talking about it again.